base/abstractarraymath.jl

```diff
@@ -330,89 +308,171 @@ julia> repeat([1 2; 3 4], inner=(2, 1), outer=(1, 3))
-_repeat_inner_outer(A::AbstractVecOrMat, ::Nothing, r::Integer) = repeat(A, r)
-
-_repeat_inner_outer(A, ::Nothing, ::Nothing) = A
-_repeat_inner_outer(A, ::Nothing, outer) = _repeat(A, ntuple(n->1, Val(ndims(A))), rep_kw2tup(outer))
-_repeat_inner_outer(A, inner, ::Nothing) = _repeat(A, rep_kw2tup(inner), ntuple(n->1, Val(ndims(A))))
-_repeat_inner_outer(A, inner, outer)     = _repeat(A, rep_kw2tup(inner), rep_kw2tup(outer))
-
-rep_kw2tup(n::Integer) = (n,)
-rep_kw2tup(v::AbstractArray{<:Integer}) = (v...,)
-rep_kw2tup(t::Tuple) = t
-
-rep_shapes(A, i, o) = _rshps((), (), size(A), i, o)
-
-_rshps(shp, shp_i, ::Tuple{}, ::Tuple{}, ::Tuple{}) = (shp, shp_i)
-@inline _rshps(shp, shp_i, ::Tuple{}, ::Tuple{}, o) =
-    _rshps((shp..., o[1]), (shp_i..., 1), (), (), tail(o))
-@inline _rshps(shp, shp_i, ::Tuple{}, i, ::Tuple{}) = (n = i[1];
-    _rshps((shp..., n), (shp_i..., n), (), tail(i), ()))
-@inline _rshps(shp, shp_i, ::Tuple{}, i, o) = (n = i[1];
-    _rshps((shp..., n * o[1]), (shp_i..., n), (), tail(i), tail(o)))
-@inline _rshps(shp, shp_i, sz, i, o) = (n = sz[1] * i[1];
-    _rshps((shp..., n * o[1]), (shp_i..., n), tail(sz), tail(i), tail(o)))
-_rshps(shp, shp_i, sz, ::Tuple{}, ::Tuple{}) =
-    (n = length(shp); N = n + length(sz); _reperr("inner", n, N))
-_rshps(shp, shp_i, sz, ::Tuple{}, o) =
-    (n = length(shp); N = n + length(sz); _reperr("inner", n, N))
-_rshps(shp, shp_i, sz, i, ::Tuple{}) =
-    (n = length(shp); N = n + length(sz); _reperr("outer", n, N))
-_reperr(s, n, N) = throw(ArgumentError("number of " * s * " repetitions " *
-    "($n) cannot be less than number of dimensions of input ($N)"))
-
-_negreperr(n) = throw(ArgumentError("number of $n repetitions" *
-    "cannot be negative"))
-
-@noinline function _repeat(A::AbstractArray, inner, outer)
-    any(<(0), inner) && _negreperr("inner")
-    any(<(0), outer) && _negreperr("outer")
-
-    shape, inner_shape = rep_shapes(A, inner, outer)
-
-    R = similar(A, shape)
-    if any(iszero, shape)
-        return R
+module _RepeatInnerOuter
+
+function repeat(arr; inner=nothing, outer=nothing)
+    check(arr, inner, outer)
+    arr, inner, outer = resolve(arr, inner, outer)
+    repeat_inner_outer(arr, inner, outer)
+end
+
+to_tuple(t::Tuple) = t
+to_tuple(x::Integer) = (x,)
+to_tuple(itr) = tuple(itr...)
+
+filltuple(val, N) = ntuple(_->val, N)
+
+function pad(a::Tuple, b::Tuple, c::Tuple)
+    a1,b1 = pad(a,b)
+    b2,c2 = pad(b1, c)
+    a2,c3  = pad(a1, c2)
+    return a2, b2, c2
+end
+function pad(a::Tuple{}, b::NTuple{N, Any}) where {N}
+    filltuple(1, Val(N)), b
+end
+function pad(a::NTuple{N, Any}, b::Tuple{}) where {N}
+    (a, filltuple(1, Val(N)))
+end
+function pad(a::NTuple{N,Any}, b::NTuple{N,Any}) where {N}
+    (a,b)
+end
+function pad(a::Tuple{}, b::Tuple{})
+    (a,b)
+end
+function pad(a::Tuple, b::Tuple)
+    a1 = first(a)
+    b1 = first(b)
+    a_tail, b_tail = pad(Base.tail(a), Base.tail(b))
+    tuple(a1, a_tail...), tuple(b1, b_tail...)
+end
```

Comment:
You can use `Base.fill_to_length` to make this much simpler:

```suggestion
function pad(a, b)
    N = max(length(a), length(b))
    Base.fill_to_length(a, 1, Val(N)), Base.fill_to_length(b, 1, Val(N))
end
function pad(a, b, c)
    N = max(max(length(a), length(b)), length(c))
    Base.fill_to_length(a, 1, Val(N)), Base.fill_to_length(b, 1, Val(N)), Base.fill_to_length(c, 1, Val(N))
end
```

Reply:
Nice!